docs/datastorage/codecs.md

```diff
@@ -0,0 +1,437 @@
+
+  public String s() { /* ... */ }
+
+  public int i() { /* ... */ }
+
+  public boolean b() { /* ... */ }
+}
+```
+
+#### Fields
+
+An `Instance` can define up to 16 fields using `#group`. Each field must be an application defining the instance the object is being made for and the type of the object. The simplest way to meet this requirement is by taking a `Codec`, setting the name of the field to decode from, and setting the getter used to encode the field.
+
+A field can be created from a `Codec` using `#fieldOf`, if the field is required, or `#optionalFieldOf`, if the field is wrapped in an `Optional` or defaulted. Either method requires a string containing the name of the field in the encoded object. The getter used to encode the field can then be set using `#forGetter`, taking in a function which given the object, returns the field data.
+
+From there, the resulting product can be applied via `#apply` to define how the instance should construct the object for the application. For ease of convenience, the grouped fields should be listed in the same order they appear in the constructor such that the function can simply be a constructor method reference.
+
+```java
+public static final Codec<SomeObject> RECORD_CODEC = RecordCodecBuilder.create(instance -> // Given an instance
+  instance.group( // Define the fields within the instance
+    Codec.STRING.fieldOf("s").forGetter(SomeObject::s), // String
+    Codec.INT.optionalFieldOf("i", 0).forGetter(SomeObject::i), // Integer, defaults to 0 if field not present
+    Codec.BOOL.fieldOf("b").forGetter(SomeObject::b) // Boolean
+  ).apply(instance, SomeObject::new) // Define how to create the object
+);
+```
+
+```js
+// Encoded SomeObject
+{
+  "s": "value",
+  "i": 5,
+  "b": false
+}
+
+// Another encoded SomeObject
+{
+  "s": "value2",
+  // i is omitted, defaults to 0
+  "b": true
+}
+```
+
+### Transformers
+
+Codecs can be transformed into equivalent, or partially equivalent, representations through mapping methods. Each mapping method takes in two functions: one to transform the current type into the new type, and one to transform the new type back to the current type. This is done through the `#xmap` function.
+
+```java
+// A class
+public class ClassA {
+
+  public ClassB toB() { /* ... */ }
+}
+
+// Another equivalent class
+public class ClassB {
+
+  public ClassA toA() { /* ... */ }
+}
+
+// Assume there is some codec A_CODEC
+public static final Codec<ClassB> B_CODEC = A_CODEC.xmap(ClassA:toB, ClassB:toA);
+```
+
+If a type is partially equivalent, meaning that there are some restrictions during conversion, there are mapping functions which return a `DataResult` which can be used to return an error state whenever an exception or invalid state is reached.
+
+A => B | B => A | Transform Method
+:---:  | :---:  | :---
+Yes    | Yes    | `#xmap`
+Yes    | No     | `#flatComapMap`
+No     | Yes    | `#comapFlatMap`
+No     | No     | `#flatXMap`
```

Comment:
This table implies (to me at least) at first glance that these methods do not allow conversion for the entries marked with `No`. Should we be concerned about this possible confusion?